autoload/clap/cache.vim

```diff
@@ -0,0 +1,51 @@
+" Author: Mark Wu <markplace@gmail.com>
+" Description: Cache API for clap.
+
+let s:save_cpo = &cpoptions
+set cpoptions&vim
+
+let s:path_separator = has('win32') ? '\' : '/'
+let s:clap_cache_dir = get(g:, 'clap_cache_dir', '')
+
+function! clap#cache#folder() abort
+  if empty(s:clap_cache_dir)
+    let user_home = $HOME
+    if has('nvim')
+      let user_cache = stdpath('cache')
+    elseif exists('$XDG_CACHE_HOME')
+      let user_cache = $XDG_CACHE_HOME
+    else
+      let user_cache = user_home . s:path_separator . '.cache'
+    endif
+    let s:clap_cache_dir = user_cache . s:path_separator . 'clap'
+  endif
+
+  if !isdirectory(s:clap_cache_dir)
+    call mkdir(s:clap_cache_dir, 'p')
+  endif
+
+  return s:clap_cache_dir
+endf
+
+function! clap#cache#file(cache_name) abort
```

Comment:
I'm thinking about whether we should provide such an API `clap#cache#location_for(provider_id, fname)` for each provider to avoid the possible naming conflict. We use `top_cache_dir\provider_id\fname` for each cached file path.

Reply:
We can just extend `clap#cache#file(cache_name)` to `clap#cache#file(provider_d, cache_name)`, It's okay for me. But, how do I know the provider_id is a valid or not?

Reply:
Or, don't care about it, just do what user typed?

Reply:
The user of this API always knows the provider id, it's generally ensured by the whole system(`autoload` and identifier rule). And `clap#cache#file` is not a general name, for people could merely want to a temp directory instead of a file.

Reply:
Let's just keep these two public APIs then.

Reply:
Done!